### PR TITLE
[WiP] OpenApi Support

### PR DESCRIPTION
**Pull request by sauterl**
Refactored Rest API to support [OpenApi](https://swagger.io/docs/specification/about/).

* Added OAS v2 support for REST API
* Added OAS live docs
* Added gradle task for OAS generation

EDIT: The result of the OAS generation gradle task kan be seen [in the Swagger Editor](https://editor.swagger.io/?url=https://raw.githubusercontent.com/vitrivr/cineast/multi-module-openapi-support/docs/swagger.json)

This leverages the work of https://github.com/manusant/spark-swagger (which was forked to [another repo](https://github.com/sauterl/spark-swagger) to quickly address some issues in the original version)

### CODE REVIEW

**Comment by silvanheller** [comment]
- Are there plans to make a PR from https://github.com/sauterl/spark-swagger to upstream?
- Does this require changes to vitrivr-ng?

File: citystories.json [hunk 1]
```diff
@@ -0,0 +1,161 @@
+{
```
Comment (silvanheller) [inline citystories.json]: I don't think this file belongs to this branch

File: citystories_remote.json [hunk 2]
```diff
@@ -0,0 +1,162 @@
+{
```
Comment (silvanheller) [inline citystories_remote.json]: Neither this file

File: docs/swagger.json [hunk 3]
```diff
@@ -0,0 +1,882 @@
...
+    "license" : {
+      "name" : "Apache 2.0",
+      "url" : "http://www.apache.org/licenses/LICENSE-2.0.html"
+    }
+  },
+  "host" : "localhost:4567",
+  "basePath" : "/",
+  "tags" : [ {
+    "name" : "api/v1",
+    "description" : "namespace description or what?"
```
Comment (silvanheller) [inline docs/swagger.json]: I think this is the tag description

**Comment by sauterl** [comment]
There were no changes on the REST API performed, hence there should(tm) be no need to do any changes on vitrivr-ng.
However, there are plans to alter the REST API (which would A. be reflected in the OAS version specification and B. lead to changes in vitrivr-ng).

Currently, rather than a PR to upstream of https://github.com/sauterl/spark-swagger the plan is to work towards support OAS v3

File: citystories.json (hunk 1 above)
Comment (sauterl) [reply citystories.json]: Thanks and done

File: docs/swagger.json (hunk 3 above)
Comment (sauterl) [reply docs/swagger.json]: The question here is what we want to have in the description in particular.
See https://github.com/vitrivr/cineast/issues/87

File: cineast-api/build.gradle [hunk 4]
```diff
@@ -17,6 +17,21 @@ distributions {
...
+    resolutionStrategy.cacheChangingModulesFor 0, 'seconds'
+}
+
+repositories {
+    maven {
+        url "https://packagecloud.io/manusant/beerRepo/maven2"
+    }
+    maven {
+        // DBIS Nexus
+        url "https://dbis-nexus.dmi.unibas.ch/repository/maven2/"
```
Comment (lucaro) [inline cineast-api/build.gradle]: Why is this needed?

File: cineast-api/src/main/java/org/vitrivr/cineast/api/APIEndpoint.java [hunk 5]
```diff
@@ -88,7 +125,14 @@ private static Service dispatchService(boolean secure) {
...
         }
 
+        /* Init Swagger */
+        if(config.getEnableLiveDoc() && swagger == null){
+            swagger = SparkSwagger.of(service).ignores(IgnoreSpec.newBuilder().withIgnoreAnnotated(
+                JsonIgnore.class)::build).endpoint(instance);
+        }
 
+/*
+        // TODO re-implement this functionality
```
Comment (lucaro) [inline cineast-api/src/main/java/org/vitrivr/cineast/api/APIEndpoint.java]: This should be rather straight forward using the content serving functionality which is also used for multimedia objects and thumbnails.

File: cineast-api/src/main/java/org/vitrivr/cineast/api/APIEndpoint.java [hunk 6]
```diff
@@ -122,66 +166,175 @@ private static Service dispatchService(boolean secure) {
...
+
+    @Override
+    public void bind(final SparkSwagger restApi){
+        ApiEndpoint apiEndpoint = registerSparkSwaggerEndpoint(restApi);
+        registerRoutesOnEndpoint(apiEndpoint);
+    }
+
+    private void registerOperations(){
+        /* Add your operations here */
+        registeredOperations.addAll(
```
Comment (lucaro) [inline cineast-api/src/main/java/org/vitrivr/cineast/api/APIEndpoint.java]: Why does a private method populate an externally defined data structure? The list should be created and returned here in order to ensure idempotence.

File: cineast-api/src/main/java/org/vitrivr/cineast/api/docs/GenerateOpenApiSpecs.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ * @author loris.sauter
+ */
+public class GenerateOpenApiSpecs {
+
+  /**
+   * Entrypoint for Cineast API application.
+   *
+   * @param args Program arguments.
+   */
+  public static void main(String[] args) {
```
Comment (lucaro) [inline cineast-api/src/main/java/org/vitrivr/cineast/api/docs/GenerateOpenApiSpecs.java]: Shouldn't this also be a runnable analogously to all cli commands?

File: cineast-api/build.gradle (hunk 4 above)
Comment (sauterl) [reply cineast-api/build.gradle]: To have our `Spark-Swagger` fork available

File: cineast-api/src/main/java/org/vitrivr/cineast/api/APIEndpoint.java (hunk 5 above)
Comment (sauterl) [reply cineast-api/src/main/java/org/vitrivr/cineast/api/APIEndpoint.java]: Thanks

File: cineast-api/build.gradle (hunk 4 above)
Comment (lucaro) [reply cineast-api/build.gradle]: As long as the build depends on internal repositories, we cannot publish it. What is part of this fork? Is it really necessary to have a fork, can't the modifications be made on the cineast side?

Comment (silvanheller) [reply cineast-api/build.gradle]: To add on luca's point, i compared the changes from your fork to master and i don't see why we should not try to get a PR merged upstream.

**Comment by lucaro** [comment]
Would it be possible to add a status indication to the title or description of this pull request? It is not clear to me if it is supposed to be complete or not.

**Comment by sauterl** [comment]
Will be closed as there is a more sophisticated solution ongoing. See https://github.com/vitrivr/cineast/pull/105
